i3562: Add option for instruction-only trace collection to drcachesim.

abhinav92003 opened the pull request:
Add option to drcachesim to include only instruction fetch entries and omit data entries from trace.

Issue: #3562

AssadHashmi commented:
Added @abhinav92003 to AArch64 pre-commit ACL.

AssadHashmi commented:
run arm tests

derekbruening commented on clients/drcachesim/common/options.cpp:
```diff
@@ -177,6 +177,12 @@ droption_t<bytesize_t> op_L0D_size(
     "Must be a power of 2 and a multiple of -line_size, unless it is set to 0, "
     "which disables data entries from appearing in the trace.");
 
+droption_t<bool> op_inst_only_trace(
```
Given that existing options involving instructions use `instr` to abbreviate it (`op_trace_after_instrs`, `op_online_instr_types`) I would vote for consistency here: so `op_instr_only_trace`.

abhinav92003 replied:
Done.

derekbruening commented on suite/tests/CMakeLists.txt:
```diff
@@ -2874,6 +2874,7 @@ endif ()
       torunonly_drcachesim(filter-simple ${ci_shared_app} "-L0_filter" "")
       torunonly_drcachesim(filter-no-i ${ci_shared_app} "-L0_filter -L0I_size 0" "")
       torunonly_drcachesim(filter-no-d ${ci_shared_app} "-L0_filter -L0D_size 0" "")
+      torunonly_drcachesim(filter-inst-only ${ci_shared_app} "-inst_only_trace" "")
```
Please also add an offline-trace test to make sure post-processing of raw traces works when there are no data references (raw2trace may have some issues).

abhinav92003 replied:
It seems that when -offline is present, the trace somehow continues to have data references. Checking.

abhinav92003 replied:
Offline trace collection is fixed now. Also, added the suggested test.

abhinav92003 commented on suite/tests/CMakeLists.txt:
```diff
@@ -2874,6 +2874,7 @@ endif ()
       torunonly_drcachesim(filter-simple ${ci_shared_app} "-L0_filter" "")
       torunonly_drcachesim(filter-no-i ${ci_shared_app} "-L0_filter -L0I_size 0" "")
       torunonly_drcachesim(filter-no-d ${ci_shared_app} "-L0_filter -L0D_size 0" "")
+      torunonly_drcachesim(filter-inst-only ${ci_shared_app} "-inst_only_trace" "")
```
Renamed it to instr-only-trace.

derekbruening commented:
>￼ abhinav92003 force-pushed the i3562-inst-only-trace branch from 4f88bdf to 0722987

Please don't force-push to shared branches: it destroys the code history and review comment history, and makes reviewing difficult by not giving the reviewer the ability to see just what has changed since the last review.  We have warnings about this on multiple wiki pages: https://github.com/DynamoRIO/dynamorio/wiki/Contributing#contributing-to-dynamorio, https://github.com/DynamoRIO/dynamorio/wiki/Code-Reviews#responding-to-review-comments-and-submitting-code-updates, https://github.com/DynamoRIO/dynamorio/wiki/Code-Reviews#submitting-a-change-as-a-non-project-member.  Yet it still happens, although it should not be done for git shared branches in general and not just our project.  Is there anywhere else we can put warnings to help new contributors realize it should not be done?

abhinav92003 commented:
> > ￼ abhinav92003 force-pushed the i3562-inst-only-trace branch from [4f88bdf](https://github.com/DynamoRIO/dynamorio/commit/4f88bdf841009f1d9d1c75f17e4932ae60b60209) to [0722987](https://github.com/DynamoRIO/dynamorio/commit/07229873332e622341eb2ae0a469af6a690f6170)
> 
> Please don't force-push to shared branches: it destroys the code history and review comment history, and makes reviewing difficult by not giving the reviewer the ability to see just what has changed since the last review. We have warnings about this on multiple wiki pages: https://github.com/DynamoRIO/dynamorio/wiki/Contributing#contributing-to-dynamorio, https://github.com/DynamoRIO/dynamorio/wiki/Code-Reviews#responding-to-review-comments-and-submitting-code-updates, https://github.com/DynamoRIO/dynamorio/wiki/Code-Reviews#submitting-a-change-as-a-non-project-member. Yet it still happens, although it should not be done for git shared branches in general and not just our project. Is there anywhere else we can put warnings to help new contributors realize it should not be done?

Sorry about that! I'll make sure to avoid this in future.

If it helps with the review: in this instance the force-push only changed the previous commit message, the code was unchanged,

derekbruening commented:
> If it helps with the review: in this instance the force-push only changed the previous commit message, the code was unchanged,

OK, though Github still does not show a button for viewing "Changes since last review", b/c the force breaks the history.

Just a heads up that I'm OOO today so may not finish a review until tomorrow.

abhinav92003 commented:
> OK, though Github still does not show a button for viewing "Changes since last review", b/c the force breaks the history.
Thanks for explaining. I'll be more careful.

> Just a heads up that I'm OOO today so may not finish a review until tomorrow.
No problem. Thanks.

derekbruening commented on clients/drcachesim/tracer/tracer.cpp:
```diff
@@ -1480,7 +1483,8 @@ init_thread_in_process(void *drcontext)
-        offline_file_type_t file_type = op_L0_filter.get_value()
+        offline_file_type_t file_type =
+            op_L0_filter.get_value() || op_instr_only_trace.get_value()
             ? OFFLINE_FILE_TYPE_FILTERED
```
I'm worried that conflating a filtered trace with a no-data trace will cause problems, as they are not the same.  A filtered trace inserts a separate instruction entry for every insruction, while a no-data trace just has the block tags and requires reconstruction of every single instruction in raw2trace.  Are we sure this ends up working properly in raw2trace?  Even if so, we may want a separate flag.

Actually I'm not really sure how a filtered no-data trace works in raw2trace: it looks like it will still go into append_memref()?  And hit the "Missing memref from predication, 0-iter repstr, or filter" case?  Is that what happens with this PR here as well (run `-verbose 4`)?

derekbruening replied:
It would be more efficient to skip the memrefs altogether.  Add a new type attribute for no data and use it to skip the memrefs?  (And set it for filtered no-data too?)

abhinav92003 replied:
> it looks like it will still go into append_memref()? And hit the "Missing memref from predication, 0-iter repstr, or filter" case?

Yes to both. For each memory instruction src+dst we still go into append_memref(), but return without actually appending any memref.


> I'm worried that conflating a filtered trace with a no-data trace will cause problems, as they are not the same. A filtered trace inserts a separate instruction entry for every insruction, while a no-data trace just has the block tags and requires reconstruction of every single instruction in raw2trace. Are we sure this ends up working properly in raw2trace? Even if so, we may want a separate flag.

By setting the file type to OFFLINE_FILE_TYPE_FILTERED, we're disabling address elision [here](https://github.com/DynamoRIO/dynamorio/blob/8551f131d97490203c2fedfbef2da9314a3ea45a/clients/drcachesim/tracer/raw2trace.h#L773), which causes the "Missing memref..." path to trigger.

> It would be more efficient to skip the memrefs altogether. Add a new type attribute for no data and use it to skip the memrefs? (And set it for filtered no-data too?)

Do you mean by adding another field in [raw2trace_thread_data_t](https://github.com/DynamoRIO/dynamorio/blob/8551f131d97490203c2fedfbef2da9314a3ea45a/clients/drcachesim/tracer/raw2trace.h#L1402) to indicate absence of memrefs?

derekbruening replied:
> > it looks like it will still go into append_memref()? And hit the "Missing memref from predication, 0-iter repstr, or filter" case?
> 
> Yes to both. For each memory instruction src+dst we still go into append_memref(), but return without actually appending any memref.
> 
> > I'm worried that conflating a filtered trace with a no-data trace will cause problems, as they are not the same. A filtered trace inserts a separate instruction entry for every insruction, while a no-data trace just has the block tags and requires reconstruction of every single instruction in raw2trace. Are we sure this ends up working properly in raw2trace? Even if so, we may want a separate flag.
> 
> By setting the file type to OFFLINE_FILE_TYPE_FILTERED, we're disabling address elision [here](https://github.com/DynamoRIO/dynamorio/blob/8551f131d97490203c2fedfbef2da9314a3ea45a/clients/drcachesim/tracer/raw2trace.h#L773), which causes the "Missing memref..." path to trigger.

Right, but the worry is that a future change to raw2trace will take an action on actual filtered traces which have a very different structure from this trace and break these.

> 
> > It would be more efficient to skip the memrefs altogether. Add a new type attribute for no data and use it to skip the memrefs? (And set it for filtered no-data too?)
> 
> Do you mean by adding another field in [raw2trace_thread_data_t](https://github.com/DynamoRIO/dynamorio/blob/8551f131d97490203c2fedfbef2da9314a3ea45a/clients/drcachesim/tracer/raw2trace.h#L1402) to indicate absence of memrefs?

No, I mean adding a new offline_file_type_t to more accurately describe this: a trace with no data refs.  And using it for filtered instr-only traces too.  Then raw2trace can make better decisions internally, and we have more accurate labels.
